be/src/exec/pipeline/sink/file_sink_operator.cpp

```diff
@@ -0,0 +1,311 @@
+    options.executor = FileSinkIOExecutor::instance();
+    _exec_queue_id = std::make_unique<bthread::ExecutionQueueId<const vectorized::ChunkPtr>>();
+    int ret = bthread::execution_queue_start<const vectorized::ChunkPtr>(_exec_queue_id.get(), &options,
+                                                                         &FileSinkBuffer::execute_io_task, this);
+    if (ret != 0) {
+        return Status::InternalError("start execution queue error");
+    }
+    return Status::OK();
+}
+
+Status FileSinkBuffer::append_chunk(RuntimeState* state, const vectorized::ChunkPtr& chunk) {
+    Status status = get_io_status();
+    if (!status.ok()) {
+        return status;
+    }
+    if (bthread::execution_queue_execute(*_exec_queue_id, chunk) != 0) {
+        return Status::InternalError("submit io task failed");
+    }
+    ++_num_pending_chunks;
+    return Status::OK();
+}
+
+bool FileSinkBuffer::need_input() {
+    return _num_pending_chunks < kExecutionQueueSizeLimit;
+}
+
+Status FileSinkBuffer::set_finishing() {
+    if (--_num_result_sinkers == 0) {
+        // when all writes are over, we add a nullptr as a special mark to trigger close
+        if (bthread::execution_queue_execute(*_exec_queue_id, nullptr) != 0) {
+            return Status::InternalError("submit task failed");
+        }
+        ++_num_pending_chunks;
+    }
+    return Status::OK();
+}
+
+bool FileSinkBuffer::is_finished() {
+    return _is_finished;
+}
+
+void FileSinkBuffer::cancel_one_sinker() {
+    _is_cancelled = true;
+    bthread::execution_queue_stop(*_exec_queue_id);
+}
+
+void FileSinkBuffer::close(RuntimeState* state) {
+    int64_t num_written_rows = 0;
+    if (_writer != nullptr) {
+        if (Status status = _writer->close(); !status.ok()) {
+            set_io_status(status);
+        }
+        num_written_rows = _writer->get_written_rows();
+        _writer.reset();
+    }
+
+    if (_sender != nullptr) {
+        auto query_statistic = std::make_shared<QueryStatistics>();
+        QueryContext* query_ctx = state->query_ctx();
+        query_statistic->add_scan_stats(query_ctx->cur_scan_rows_num(), query_ctx->get_scan_bytes());
+        query_statistic->add_cpu_costs(query_ctx->cpu_cost());
+        query_statistic->add_mem_costs(query_ctx->mem_cost_bytes());
+        query_statistic->set_returned_rows(num_written_rows);
+        _sender->set_query_statistics(query_statistic);
+        Status final_status = _fragment_ctx->final_status();
+        Status io_status = get_io_status();
+        if (!io_status.ok() && final_status.ok()) {
+            final_status = io_status;
+        }
+        _sender->close(final_status);
+        _sender.reset();
+
+        _state->exec_env()->result_mgr()->cancel_at_time(time(nullptr) + config::result_buffer_cancelled_interval_time,
+                                                         state->fragment_instance_id());
+    }
+    _is_finished = true;
+    bthread::execution_queue_stop(*_exec_queue_id);
```

Comment:
what will happen if `RETURN_IF_ERROR(_writer->init(state));` failed ?

Reply:
> what will happen if `RETURN_IF_ERROR(_writer->init(state));` failed ?

`_writer->init(state)`? this is invoked in prepare stage, if it failed, the query will return an error

Reply:
> what will happen if `RETURN_IF_ERROR(_writer->init(state));` failed ?

fixed